rockcraft/oci.py

```diff
@@ -355,19 +355,35 @@ def set_default_user(self, user: str) -> None:
         _config_image(image_path, params)
         emit.progress(f"Default user set to {user}")
 
-    def set_entrypoint(self) -> None:
-        """Set the OCI image entrypoint. It is always Pebble and CMD is null."""
+    def set_entrypoint(self, entrypoint_service: Optional[str]) -> None:
+        """Set the OCI image entrypoint. It is always Pebble."""
         emit.progress("Configuring entrypoint...")
         image_path = self.path / self.image_name
         entrypoint = [f"/{Pebble.PEBBLE_BINARY_PATH}", "enter", "--verbose"]
+        if entrypoint_service:
+            entrypoint.extend(["--args", entrypoint_service])
         params = ["--clear=config.entrypoint"]
         for entry in entrypoint:
             params.extend(["--config.entrypoint", entry])
         _config_image(image_path, params)
-        # Clear the CMD
-        _config_image(image_path, ["--clear=config.cmd"])
         emit.progress(f"Entrypoint set to {entrypoint}")
 
+    def set_cmd(self, command: Optional[str]) -> None:
+        """Set the OCI image CMD."""
+        emit.progress("Configuring CMD...")
+        image_path = self.path / self.image_name
+        cmd_params = ["--clear=config.cmd"]
+        if command:
+            parts = command.split()
+        else:
+            parts = []
+        for i in range(1, len(parts)):
+            if parts[i] != "[" and parts[i] != "]":
+                cmd_params.extend(["--config.cmd", parts[i]])
```

Comment:
only noticing this now, which means **we need another test**.

I'm 90% this won't work for most cases. basically, you're assuming that a service command will always be something like `command: app [ args ]`, but in reality, it can be `command: app -v --fixed-option --arg foo [ --some extra-arg --etc ]`.

This means:
 1. we need unit tests for assessing more variations of the service command
 2. this block of code needs to be more generic.

For 2. I suggest doing something like:

```python
# up top
import shlex
#...
def set_cmd(self, command: Optional[str]) -> None:
    """Set the OCI image CMD."""
    emit.progress("Configuring CMD...")
    image_path = self.path / self.image_name
    cmd_params = ["--clear=config.cmd"]
    command_sh_args = shlex.split(command)
    try:
        opt_args = command_sh_args[command_sh_args.index("[")+1:command_sh_args.index("]")]
    except ValueError:
        emit.debug(f"The entrypoint-service command '{command}' has no default arguments. CMD won't be set."
        return
    cmd_params.extend(opt_args)
    _config_image(image_path, cmd_params)
```
(haven't tested)

P.S.: why is `shlex` needed? Well, you must consider the following (which btw, is yet another unit test we need for this): `command: app --arg "[ foo ]" [ --optional ]`. IOW, if passed as a string, the squared brackets have no effect in pebble, and are simply treated as any other argument for the command app.